### PR TITLE
Node Agent Integration

### PR DESCRIPTION
**Pull request by mmoreiradj**
## gRPC Worklod Service Server Integration

Add a wrapper for `containerd-client` lib (and `ctr` cli for creation) to create a container and run a workload.

The wrapper uses gRPC to communicate to the containerd socket. It reconnects to the socket before each request. This is done to avoid potential issues with resource management, isolation, and security and simplicity. It is also more fault-tolerant.

This will however mean more resource usage, overhead, latency, potential socket exhaustion, etc.. This is a tradeoff we are willing to make for now, since these issues are not likely to be a problem in the near future.

The `ctr` cli was used to create the container because the `containerd-client` we did not manage to create a container with the `containerd-client` lib. It runs the command `ctr run -d <image> <container-id>`.

The `containerd-client` lib was also used to kill the container. We did however have an issue concerning an ambiguous type efinition. Indeed, the Workload Signal enum is ambiguous. It is not clear what signal STOP represents. For now we made the assumption that stop is SIGINT (2). Both Stop (SIGINT) and Kill (SIGKILL) signals are sent to the workload, then the workload is cleaned up.

**Note**: It seems Kuruyia has managed to prepare a rootfs and run a container, you can find it [here](https://github.com/Martin-Moreira-de-jesus/orka/blob/tmp/containerd-create-workload/node-agent/src/workload_manager/container/client.rs).

But to be honest, going with `runc` or a container runtime made in rust might be an easier way to achieve our goals.

## gRPC Lifecycle Service Client and Status Update Service Client Integration

When the program starts, it connects to the scheduler and registers itself to it. If it fails, it will retry every 5 seconds. If it fails 3 times (can be modified in args or env see `node-agent/src/args.rs`), it will exit. It also starts the gRPC server.

If any of the two processes fail (gRPC of lifecycle), the server will shut down and try to leave the cluster if needed.

Note: We would be glad to know how you think we managed the application's lifecycle. It was a first for us, and it feels clunky.

## Cargo Workspaces

We added a `Cargo.toml` file at the root of the project to manage the dependencies of the different crates. Building, formatting and linting will now be easier.

However, we had an issue when building the node-agent. When you build the binary from the root of the project, requests to the `containerd` socket fails, saying it didn't find it. We did not manage to fix this issue. We did however manage to make it work only if build from the node-agent directory. We are not sure why this is happening.

## Misc

Add ide directory to gitignore.

## What's next

Use the network team's CNI implementation.

@sameo 
@noetarbouriech

### CODE REVIEW

File: node-agent/workload-manager/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+[package]
+name = "workload-manager"
+version = "0.1.0"
+edition = "2021"
+
+[dependencies]
+anyhow = "1.0.72"
+containerd-client = {git = "https://github.com/containerd/rust-extensions", branch="main"}
```
Comment (sameo) [inline node-agent/workload-manager/Cargo.toml]: FWIW, interacting at the containerd level seems like a reasonable approach.

**Comment by sameo** [comment]
@Martin-Moreira-de-jesus Let me know when I should review this PR, it is still in draft.

**Comment by mmoreiradj** [comment]
Sure, we've been struggling with containerd so far. We're getting there...

**Comment by sameo** [comment]
To be rebased.

**Comment by mmoreiradj** [comment]
> .

@sameo done !

**Comment by mmoreiradj** [comment]
I just realized cargo fmt formatted the controller's code, is that an issue ?

**Pull request closed without merging**